### ISSUE TITLE
[bus-mapping] Use a trait Object to avoid repeating the call resolving `OpcodeId` to `Opcode`

### ISSUE THREAD

**Issue by CPerezz** [appliedzkp/zkevm-circuits#138]
To provide some context, this refers to the situation that happens when we need to downcast from `OpcodeId` into it's `Opcode` trait repr: [here](https://github.com/appliedzkp/zkevm-circuits/blob/main/bus-mapping/src/evm/opcodes.rs#L40-L50)
```rust
// This is implemented for OpcodeId so that we can downcast the responsabilities
// to the specific Opcode structure implementations since OpcodeId is a single
// structure with all the OPCODES stated as associated constants.
// Easier to solve with a macro. But leaving here for now until we refactor in a
// future PR.
impl Opcode for OpcodeId {
    fn gen_associated_ops(
        &self,
        ctx: &mut TraceContext,
        exec_step: &mut ExecutionStep,
        next_steps: &[ExecutionStep],
    ) -> Result<(), Error> {
        match *self {
            OpcodeId::PUSH1 => {
                Push1 {}.gen_associated_ops(ctx, exec_step, next_steps)
            }
            OpcodeId::MLOAD => {
                Mload {}.gen_associated_ops(ctx, exec_step, next_steps)
            }
            OpcodeId::SLOAD => {
                Sload {}.gen_associated_ops(ctx, exec_step, next_steps)
            }
            OpcodeId::STOP => {
                Stop {}.gen_associated_ops(ctx, exec_step, next_steps)
            }
            _ => unimplemented!(),
        }
    }
}
```

We need a better solution that an enormous match here.

I think for sure we should find a way to avoid having the same call with the same arguments for each case, we just need to resolve an `OpcodeId` to `Opcode`.  I think a macro is not the answer here. 

_Originally posted by @ed255 in https://github.com/appliedzkp/zkevm-circuits/pull/134#discussion_r734973900_

**Comment by ed255** [comment]
In https://github.com/appliedzkp/zkevm-circuits/pull/162 I addressed this by introducing a function with a match with one case for each opcode and returning the gen_associated_ops function, so that we don't have to repeat the call with the same arguments over and over again: https://github.com/appliedzkp/zkevm-circuits/blob/80666a55b43eaa6ad09d4bc83494f7870c78cf5f/bus-mapping/src/evm/opcodes.rs#L41-L50

I think we can't get rid of the match, and I think that's OK!

**Comment by ed255** [comment]
Closing as this issue has been resolved in #162
